src/NHibernate/Transaction/AdoNetWithDistributedTransactionFactory.cs

```diff
@@ -94,90 +79,196 @@ public class DistributedTransactionContext : ITransactionContext, IEnlistmentNot
 		{
 			public System.Transactions.Transaction AmbientTransation { get; set; }
 			public bool ShouldCloseSessionOnDistributedTransactionCompleted { get; set; }
-			private readonly ISessionImplementor sessionImplementor;
+
+			private readonly ISessionImplementor _sessionImplementor;
+			private readonly ManualResetEvent _waitEvent = new ManualResetEvent(true);
+			private readonly AsyncLocal<bool> _bypassWait = new AsyncLocal<bool>();
+
 			public bool IsInActiveTransaction;
 
-			public DistributedTransactionContext(ISessionImplementor sessionImplementor, System.Transactions.Transaction transaction)
+			public DistributedTransactionContext(
+				ISessionImplementor sessionImplementor,
+				System.Transactions.Transaction transaction)
 			{
-				this.sessionImplementor = sessionImplementor;
+				_sessionImplementor = sessionImplementor;
 				AmbientTransation = transaction.Clone();
 				IsInActiveTransaction = true;
 			}
 
+			public void WaitOne()
+			{
+				if (_bypassWait.Value || _isDisposed)
+					return;
+				try
+				{
+					if (!_waitEvent.WaitOne(5000))
+					{
+						// A call occurring after transaction scope disposal should not have to wait long, since
+						// the scope disposal is supposed to block until the transaction has completed: I hope
+						// that it at least ensure IO are done, even if experience shows DTC lets the scope
+						// disposal leave before having finished with volatile ressources and
+						// TransactionCompleted event.
+						_waitEvent.Set();
+						throw new HibernateException(
+							"Synchronization timeout for transaction completion. This is very likely a bug in NHibernate.");
+					}
+				}
+				catch(Exception ex)
+				{
+					_logger.Warn(
+						"Synchronization failure, assuming it has been concurrently disposed and do not need sync anymore.",
+						ex);
```

Comment:
I have not witnessed this, but it may happen if between the test on `_isDisposed` and the call to `WaitOne`, the transaction completion event dispose the wait event.